File: android/src/main/kotlin/com/ko2ic/fluttergoogleadmanager/RewardedAd.kt
package com.ko2ic.fluttergoogleadmanager

import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.doubleclick.PublisherAdRequest
import com.google.android.gms.ads.reward.RewardItem
import com.google.android.gms.ads.reward.RewardedVideoAdListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class RewardedAd(private val registrar: PluginRegistry.Registrar, private val channel: MethodChannel) :
    MethodChannel.MethodCallHandler {

    private val rewardedAd = MobileAds.getRewardedVideoAdInstance(registrar.context())

    init {
        rewardedAd.rewardedVideoAdListener = RewardedAdListener(channel);
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "load" -> load(call, result)
            "show" -> show(result)
            "dispose" -> dispose(result)
            "pause" -> rewardedAd.pause(registrar.context())
            "resume" -> rewardedAd.resume(registrar.context())
            else -> result.notImplemented()
        }
    }

    private fun load(call: MethodCall, result: MethodChannel.Result) {
        val isDevelop = call.argument<Boolean>("isDevelop") ?: false

        if (isDevelop) {
            rewardedAd.loadAd("/6499/example/rewarded-video", PublisherAdRequest.Builder().build())
        } else {
            rewardedAd.loadAd(call.argument<String>("adUnitId"), PublisherAdRequest.Builder().build())
        }
        result.success(null)
    }

    private fun show(result: MethodChannel.Result) {
        if (rewardedAd.isLoaded) {
            rewardedAd.show()
            result.success(null)
        } else {
            result.error("not_loaded_yet", "The Rewarded Ads wasn't loaded yet", null)
        }
    }

    private fun dispose(result: MethodChannel.Result) {
        rewardedAd.destroy(registrar.context())
        result.success(null)
    }

    class RewardedAdListener(private val channel: MethodChannel) : RewardedVideoAdListener {
        override fun onRewarded(reward: RewardItem) {
            channel.invokeMethod("onRewarded", mapOf("type" to reward.type, "amount" to reward.amount))
        }

        override fun onRewardedVideoAdLeftApplication() {
            channel.invokeMethod("onAdLeftApplication", null)
        }

        override fun onRewardedVideoAdClosed() {
            channel.invokeMethod("onAdClosed", null)
        }

        override fun onRewardedVideoAdFailedToLoad(errorCode: Int) {
            channel.invokeMethod("onAdFailedToLoad", mapOf("errorCode" to errorCode))
        }

        override fun onRewardedVideoAdLoaded() {
            channel.invokeMethod("onAdLoaded", null)
        }

        override fun onRewardedVideoAdOpened() {
            channel.invokeMethod("onAdOpened", null)
        }

        override fun onRewardedVideoStarted() {
            channel.invokeMethod("onVideoStarted", null)
        }

        override fun onRewardedVideoCompleted() {
            channel.invokeMethod("onVideoCompleted", null)
        }
    }
}

File: ios/Classes/RewardedAd.swift
import Flutter
import GoogleMobileAds

class RewardedAd: SwiftFlutterGoogleAdManagerPlugin {
    private let channel: FlutterMethodChannel!

    /// The reward-based video ad.
    weak var rewardedAd: GADRewardedAd?

    private let exampleReward = "/6499/example/rewarded-video"

    public init(with _: FlutterPluginRegistrar, channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
        channel.setMethodCallHandler({
            (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            self.handle(call, result: result)
        })
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load":
            load(call, result: result)
        case "show":
            show(call, result: result)
        case "dispose":
            dispose(call, result: result)
        case "pause":
            break
        case "resume":
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func load(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let argument = call.arguments as! Dictionary<String, Any>
        let isDevelop = argument["isDevelop"] as? Bool ?? false


        let unitId: String? = isDevelop ? exampleReward : argument["adUnitId"] as? String
        GADRewardedAd.load(
            withAdUnitID: unitId ?? "", request: GAMRequest()
        ) { (ad, error) in
            if let error = error {
                print("Rewarded ad failed to load with error: \(error.localizedDescription)")
                return
            }
            print("Loading Succeeded")
            self.rewardedAd = ad
            self.rewardedAd?.fullScreenContentDelegate = self
        }
        result(nil)
    }


    private func show(_: FlutterMethodCall, result: @escaping FlutterResult) {

        guard let ad = rewardedAd else {
            result(FlutterError.notLoad)
            return
        }

        guard let rootViewController = UIApplication.rootViewController else {
            result(FlutterError.controllerError)
            return
        }

        ad.present(fromRootViewController: rootViewController) { [weak self] in
            guard let self = self else { return }

            let reward = ad.adReward
            self.channel.invokeMethod("onRewarded", arguments: ["type": reward.type, "amount": reward.amount])
            result(nil)
        }
    }

    private func dispose(_: FlutterMethodCall, result: @escaping FlutterResult) {
        result(nil)
    }
}



extension RewardedAd: GADFullScreenContentDelegate {
    func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        // TODO:
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
//       TODO:
    }

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        channel.invokeMethod("onAdFailedToLoad", arguments: ["errorCode": (error as NSError).code])
    }

}


//
//extension RewardedAd: GADRewardBasedVideoAdDelegate {
//    func rewardBasedVideoAd(_: GADRewardBasedVideoAd,
//                            didRewardUserWith reward: GADAdReward) {
//        channel.invokeMethod("onRewarded", arguments: ["yype": reward.type, "amount": reward.amount])
//    }
//
//    func rewardBasedVideoAdDidReceive(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onAdLoaded", arguments: nil)
//    }
//
//    func rewardBasedVideoAdDidOpen(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onAdOpened", arguments: nil)
//    }
//
//    func rewardBasedVideoAdDidStartPlaying(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onVideoStarted", arguments: nil)
//    }
//
//    func rewardBasedVideoAdDidCompletePlaying(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onVideoCompleted", arguments: nil)
//    }
//
//    func rewardBasedVideoAdDidClose(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onAdClosed", arguments: nil)
//    }
//
//    func rewardBasedVideoAdWillLeaveApplication(_: GADRewardBasedVideoAd) {
//        channel.invokeMethod("onAdLeftApplication", arguments: nil)
//    }
//
//    func rewardBasedVideoAd(_: GADRewardBasedVideoAd,
//                            didFailToLoadWithError error: Error) {
//        channel.invokeMethod("onAdFailedToLoad", arguments: ["errorCode": (error as NSError).code])
//    }
//}
